sklearn/metrics/base.py

```diff
@@ -131,3 +132,94 @@ def _average_binary_score(binary_metric, y_true, y_score, average,
     else:
         return score
+
+def _average_multiclass_score(binary_metric, y_true, y_score,
+                              average, multiclass):
+
+    """Uses the binary metric for multiclass classification
+
+    Parameters
+    ----------
+    y_true : array, shape = [n_samples] or [n_samples, n_classes]
+        True multiclass labels
+
+    y_score : array, shape = [n_samples] or [n_samples, n_classes]
+        Target scores corresponding to probability estimates of a sample
+        belonging to a particular class
+
+    average : string, [None, 'macro' (default), 'weighted']
+        TODO: difference between 'macro' and None? Should there be both?
+        If ``None``, the scores for each class are returned. Otherwise,
+        this determines the type of averaging performed on the data:
+
+        ``'macro'``:
+            Calculate metrics for each label, and find their unweighted
+            mean.  This does not take label imbalance into account.
+        ``'weighted'``:
+            Calculate metrics for each label, taking into account the a priori
+            distribution of the classes.
+
+    binary_metric : callable, returns shape [n_classes]
+        The binary metric function to use.
+
+    Returns
+    -------
+    score : float or array of shape [n_classes]
+        If not ``None``, average the score, else return the score for each
+        classes.
+
+    """
+    average_options = (None, "macro", "weighted")
+    if average not in average_options:
+        raise ValueError("average has to be one of {0}"
+                         "".format(average_options))
+    multiclass_options = ("ovo", "ovr")
+    if multiclass not in multiclass_options:
+        raise ValueError("{0} is not supported for multiclass ROC AUC"
+                         "".format(multiclass))
+
+    check_consistent_length(y_true, y_score)
+    y_true = check_array(y_true)
+    y_score = check_array(y_score)
+
+    not_average_axis = 1
+
+    if y_true.ndim == 1:
+        y_true = y_true.reshape((-1, 1))
+
+    if y_score.ndim == 1:
+        y_score = y_score.reshape((-1, 1))
+
+    label_unique, label_counts = np.unique(y_true, return_counts=True)
+    label_counts_map = dict(zip(label_unique, label_counts))
+    n_labels = len(label_unique)
+    if multiclass == "ovo":
+        # Hand and Till 2001 (unweighted)
+        pairwise = [p for p in itertools.combinations(xrange(n_labels), 2)]
+        auc_scores_sum = 0
+        for pair in pairwise:
+            ix = np.in1d(y_true.ravel(), [pair[0], pair[1]])
+            y_true_filtered = y_true[0, np.where(ix)]
+            y_score_filtered = y_score[np.where(ix)]
+            y_true_filtered_10 = np.in1d(y_true_filtered.ravel(), pair[0]).astype(int)
+            y_true_filtered_01 = np.in1d(y_true_filtered.ravel(), pair[1]).astype(int)
+            binary_avg_output = \
+              (binary_metric(y_true_filtered_10, y_score_filtered[:,pair[0]]) +
```

Comment:
In the paper, Hand & Till define A(i, j) as [A(i | j) + A(j | i)] / 2 [Middle of page 177 on [here](http://link.springer.com/article/10.1023/A:1010920819831), a few lines above equation (7)].
My implementation was based on this paper, though I may revise to follow the "AU1U" in the [paper I referred to](https://www.math.ucdavis.edu/~saito/data/roc/ferri-class-perf-metrics.pdf) in the other comment about doing weighted OvO.